Presentation docs contentinstancesofspecificclasses (#2979)
iTwin/itwinjs-core: docs/presentation/Content/ContentInstancesOfSpecificClasses.md

# Content Instances Of Specific Classes Specification

> TypeScript type: [ContentInstancesOfSpecificClassesSpecification]($presentation-common).

> A specification that creates content for  instances of specific ECClasses.

Reviewer: The problem that I see with **all** is that it's kind of a lie - as you said yourself, it's not *always* and in practice I think it's actually a rare case to use this specification without any additional filtering, in which case it doesn't load all instances. Maybe an additional sentence like "All instances of requested classes are loaded unless additional filtering is used" could be added.

> A specification that creates content for  instances of specific ECClasses.

Author: I prefer @roluk version since, unless specified in the filter, it does select all instances of the target classes.
@grigasp mentioned that it's usually used with some filter, but the filter is not mandatory so I don't think adding `"All instances of requested classes are loaded unless additional filtering is used"` is very accurate either.

| [`propertyOverrides`](#attribute-propertyoverrides)                             | No        | `PropertySpecification[]`                                                                                                    | `[]`    |
| [`showImages`](#attribute-showimages)                                           | No        | `boolean`                                                                                                                    | `false` |
| *Misc.*                                                                         |
| [`relatedInstances`](#attribute-relatedinstances)                               | No        | [`RelatedInstanceSpecification[]`](../Common-Rules/RelatedInstanceSpecification.md)                                          | `[]`    |

### Attribute: `classes`

Defines a single or an array of [multi schema classes](../Common-Rules/MultiSchemaClassesSpecification.md) which specify what and how the classes need to be selected to form result content.

```ts

> [[include:ContentInstancesOfSpecificClasses.Classes.Ruleset]]

Author: The instances are from Model class which, as far as I've understood, don't have many properties since those are located in the elements that the model represents.

> [[include:ContentInstancesOfSpecificClasses.Classes.Ruleset]]

Reviewer: Correct, the base `Model` class has only 1 visible property.

  | with filter    | ![Example when filtering instances](./media/contentinstancesofspecificclasses-with-instancefilter-2.png)     |

### Attribute: `onlyIfNotHandled`

> **Default value:** `false`

Identifies whether we should ignore this specification if another specification was already handled (based on rule priorities and definition order). Should be used when defining a fallback specification.

```ts
[[include:ContentInstancesOfSpecificClasses.OnlyIfNotHandled.Ruleset]]
```

  | onlyIfNotHandled | Result                                                                                                         |
  | ---------------- | -------------------------------------------------------------------------------------------------------------- |
  | `true`           | ![Example when selecting all instances](./media/contentinstancesofspecificclasses-with-onlyifnothandled-1.png) |
  | `false`          | ![Example when filtering instances](./media/contentinstancesofspecificclasses-with-onlyifnothandled-2.png)     |

### Attribute: `priority`

> **Default value:** `1000`

> Defines the order in which specifications are handled - higher priority means the specifications is handled first. If priorities are equal, the specifications are handled in the order they're defined.

Reviewer 2: ```suggestion
Controls the order in which specifications are handled — specification with higher priority value is handled first. If priorities are equal, the specifications are handled in the order they appear in the ruleset.
```

# Content Instances Of Specific Classes Specification

> TypeScript type: [ContentInstancesOfSpecificClassesSpecification]($presentation-common).

A specification that creates content for  instances of specific ECClasses.

## Attributes

| Name                                                                            | Required? | Type                                                                                                                         | Default |
| ------------------------------------------------------------------------------- | --------- | ---------------------------------------------------------------------------------------------------------------------------- | ------- |
| *Filtering*                                                                     |
| [`classes`](#attribute-classes)                                                 | Yes       | [`MultiSchemaClassesSpecification \| MultiSchemaClassesSpecification[]`](../Common-Rules/MultiSchemaClassesSpecification.md) | `[]`    |

> | [`handleInstancesPolymorphically`](#attribute-handleinstancespolymorphically)   | No        | `boolean`                                                                                                                    | `false` |

Reviewer: I'd say we shouldn't include deprecated attributes in this table.

  | `false`          | ![Example when filtering instances](./media/contentinstancesofspecificclasses-with-onlyifnothandled-2.png)     |

### Attribute: `priority`

> **Default value:** `1000`

Defines the order in which specifications are handled - higher priority means the specifications is handled first. If priorities are equal, the specifications are handled in the order they're defined.

```ts
[[include:ContentInstancesOfSpecificClasses.Priority.Ruleset]]
```

![Example of using priority attribute](./media/contentinstancesofspecificclasses-with-priority.png)

### Attribute: `relatedProperties`

Specifications of [related properties](./RelatedPropertiesSpecification.md) which are included in the generated content.

```ts
[[include:ContentInstancesOfSpecificClasses.RelatedProperties.Ruleset]]
```

  |         | Result                                                                                                                    |
  | ------- | ------------------------------------------------------------------------------------------------------------------------- |

>   | without | ![Example when doing normal property select](./media/contentinstancesofspecificclasses-with-relatedproperties-1.png)      |

Reviewer: I think the example would look much better in property grid - it would show the actual related properties rather than a clickable link that we see in tables.

  | with    | ![Example when selecting with related properties](./media/contentinstancesofspecificclasses-with-relatedproperties-2.png) |

### Attribute: `calculatedProperties`

Specifications of [calculated properties](./CalculatedPropertiesSpecification.md) whose values are generated using provided [ECExpressions](../Advanced/ECExpressions.md#ecinstance).

```ts
[[include:ContentInstancesOfSpecificClasses.CalculatedProperties.Ruleset]]
```

![Example of using calculatedProperties attribute](./media/contentinstancesofspecificclasses-with-calculatedproperties.png)

### Attribute: `propertyCategories`

Specifications of [custom categories](PropertyCategorySpecification.md).

Simply defining the categories does nothing - they have to be referenced through [`PropertySpecification.categoryId`](./PropertySpecification.md) specified in [`propertyOverrides`](#attribute-propertyoverrides) list.

```ts
[[include:ContentInstancesOfSpecificClasses.PropertyCategories.Ruleset]]
```

> ![Example of using propertyCategories attribute](./media/contentinstancesofspecificclasses-with-propertycategories.png)

Reviewer: Problem with the screenshot is that it shows property grid content for multiple elements and the values get merged. I think we need to produce a case where content is displayed for just 1 element.

Author: what if I take example where all three instances share the same values? Filtering to take 1 element with a specification that is designed to take a whole class of elements also feels like a bad example (I would even argue that the whole attribute feels misplaced on specifications which return many instances).

| [`propertyOverrides`](#attribute-propertyoverrides)                             | No        | `PropertySpecification[]`                                                                                                    | `[]`    |
| [`showImages`](#attribute-showimages)                                           | No        | `boolean`                                                                                                                    | `false` |
| *Misc.*                                                                         |
| [`relatedInstances`](#attribute-relatedinstances)                               | No        | [`RelatedInstanceSpecification[]`](../Common-Rules/RelatedInstanceSpecification.md)                                          | `[]`    |

### Attribute: `classes`

Defines a set of [multi schema classes](../Common-Rules/MultiSchemaClassesSpecification.md) that specify which ECClasses need to be selected to form the result.

```ts
[[include:ContentInstancesOfSpecificClasses.Classes.Ruleset]]
```

![Example of using "classes" attribute](./media/contentinstancesofspecificclasses-with-classes.png)

### Attribute: `excludedClasses`

> Defines a set of [multi schema classes](../Common-Rules/MultiSchemaClassesSpecification.md) that specify which ECClasses need to be excluded from the [`classes` attribute](#attribute-classes) result.

Reviewer 2: ```suggestion
Defines a set of [multi schema classes](../Common-Rules/MultiSchemaClassesSpecification.md) that prevents specified ECClasses and subclasses from being selected by [`classes` attribute](#attribute-classes).
```

> TypeScript type: [ContentInstancesOfSpecificClassesSpecification]($presentation-common).

> Specification which allows targeting content instances based on their ECClass.

Author: for ln 1: add spaces

> Specification which allows targeting content instances based on their ECClass.

Author: for -> from - "A specification that creates content from instances of specific ECClasses."?

> Specification which allows targeting content instances based on their ECClass.

Reviewer: "from" suggests we're creating content from a single instance. However, we're creating it for a single instance and it may be related to other instances that we're going to include as well. So I'd say we're creating content for a single instance from possibly multiple instances :)

| [`propertyCategories`](#attribute-propertycategories)                           | No        | `PropertyCategorySpecification[]`                                                                                             | `[]`    |
| [`propertyOverrides`](#attribute-propertyoverrides)                             | No        | `PropertySpecification[]`                                                                                                     | `[]`    |
| [`showImages`](#attribute-showimages)                                           | No        | `boolean`                                                                                                                     | `false` |
| *Misc.*                                                                         |
| [`relatedInstances`](#attribute-relatedinstances)                               | No        | [`RelatedInstanceSpecification[]`](../Common-Rules/RelatedInstanceSpecification.md)                                           | `[]`    |

### Attribute: `classes`

> Defines a single or an array of [multi schema classes](../Common-Rules/MultiSchemaClassesSpecification.md) which will specify what and how the classes need to be selected to form result content.

Reviewer: ```suggestion
Defines a single or an array of [multi schema classes](../Common-Rules/MultiSchemaClassesSpecification.md) which specify what and how the classes need to be selected to form result content.
```

```ts
[[include:ContentInstancesOfSpecificClasses.Classes.Ruleset]]
```

![Example of using classes attribute](./media/contentInstancesOfSpecificClasses-with-classes-attribute.png)

### Attribute: `handlePropertiesPolymorphically`

> **Default value:** `false`

> Defines the default value whether the properties of derived `classes` should be included in the content.

Reviewer: ```suggestion
Specifies whether properties of derived `classes` should be included in the content.
```

```ts
[[include:ContentInstancesOfSpecificClasses.Classes.Ruleset]]
```

  | handlePropertiesPolymorphically | Result                                                                                                                                      |
  | ------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------- |
  | `false`  | ![Example when only selecting class specified properties](./media/contentInstancesOfSpecificClasses-with-handlePropertiesPolymorphically-attribute-1.png) |
  | `true`    | ![Example when selecting parent and child class properties](./media/contentInstancesOfSpecificClasses-with-handlePropertiesPolymorphically-attribute-2.png)     |

### Attribute: `instanceFilter`

> Allows defining additional conditions when filtering result target class instances. More information found [here](./ECExpressions.md#instance-filter).

Reviewer: ```suggestion
Specifies an [ECExpression](./ECExpressions.md#instance-filter) for filtering instances of ECClasses specified through the [`classes` attribute](#attribute-classes).
```

```ts
[[include:ContentInstancesOfSpecificClasses.InstanceFilter.Ruleset]]
```

  |  | Result                                                                                                                                      |
  | ------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------- |
  | without filter | ![Example when selecting all instances](./media/contentInstancesOfSpecificClasses-with-instanceFilter-attribute-1.png) |
  | with filter  | ![Example when filtering instances](./media/contentInstancesOfSpecificClasses-with-instanceFilter-attribute-2.png)     |

### Attribute: `onlyIfNotHandled`

> **Default value:** `false`

Identifies whether we should ignore this specification if another specification was already handled (based on rule priorities and definition order). Should be used when defining a fallback specification.

```ts
[[include:ContentInstancesOfSpecificClasses.OnlyIfNotHandled.Ruleset]]
```

>   | onlyIfNotHandled | Result                                                                                                                                      |
>   | ------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------- |
>   | true  | ![Example when selecting all instances](./media/contentInstancesOfSpecificClasses-with-onlyIfNotHandled-attribute-1.png) |
>   | false    | ![Example when filtering instances](./media/contentInstancesOfSpecificClasses-with-onlyIfNotHandled-attribute-2.png)     |

Reviewer: ```suggestion
  | onlyIfNotHandled | Result                                                                                                                                      |
  | ------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------- |
  | `true`  | ![Example when selecting all instances](./media/contentInstancesOfSpecificClasses-with-onlyIfNotHandled-attribute-1.png) |
  | `false`    | ![Example when filtering instances](./media/contentInstancesOfSpecificClasses-with-onlyIfNotHandled-attribute-2.png)     |
```

### Attribute: `priority`

> **Default value:** `1000`

Defines the order in which specifications are handled - higher priority means the specifications is handled first. If priorities are equal, the specifications are handled in the order they're defined.

```ts
[[include:ContentInstancesOfSpecificClasses.Priority.Ruleset]]
```

![Example of using priority attribute](./media/contentInstancesOfSpecificClasses-with-priority-attribute.png)

### Attribute: `relatedProperties`

> Specifications of [related properties](./Terminology.md#related-properties) which are included in the generated content. *See [this page](./RelatedPropertiesSpecification.md) for more details.*

Reviewer: ```suggestion
Specifications of [related properties](./RelatedPropertiesSpecification.md) which are included in the generated content.
```

```ts
[[include:ContentInstancesOfSpecificClasses.RelatedProperties.Ruleset]]
```

  |  | Result                                                                                                                                      |
  | ------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------- |
  | without  | ![Example when doing normal property select](./media/contentInstancesOfSpecificClasses-with-relatedProperties-attribute-1.png) |
  | with    | ![Example when selecting with related properties](./media/contentInstancesOfSpecificClasses-with-relatedProperties-attribute-2.png)     |

### Attribute: `calculatedProperties`

> Specifications of calculated properties whose values are generated using provided ECExpressions. *See [this page](./CalculatedPropertiesSpecification.md) for more details.*

Reviewer: ```suggestion
Specifications of [calculated properties](./CalculatedPropertiesSpecification.md) whose values are generated using provided [ECExpressions](../Advanced/ECExpressions.md#ecinstance).
```

```ts
[[include:ContentInstancesOfSpecificClasses.CalculatedProperties.Ruleset]]
```

![Example of using calculatedProperties attribute](./media/contentInstancesOfSpecificClasses-with-calculatedProperties-attribute.png)

### Attribute: `propertyCategories`

> Specifications for custom categories. Simply defining the categories does nothing - they have to be referenced from `PropertySpecification` defined in `propertyOverrides` by `id`. *See [this page](./PropertyCategorySpecification.md) for more details.*

Reviewer: ```suggestion
Specifications of [custom categories](PropertyCategorySpecification.md). 

Simply defining the categories does nothing - they have to be referenced through [`PropertySpecification.categoryId`](./PropertySpecification.md) specified in [`propertyOverrides`](#attribute-property-overrides) list.
```

```ts
[[include:ContentInstancesOfSpecificClasses.PropertyCategories.Ruleset]]
```

![Example of using propertyCategories attribute](./media/contentInstancesOfSpecificClasses-with-propertyCategories-attribute.png)

### Attribute: `propertyOverrides`

> Specifications for various property overrides. *See [this page](./PropertySpecification.md) for more details.*

Reviewer: ```suggestion
Specifications of various [property overrides](./PropertySpecification.md) that allow customizing individual properties display.
```

```ts
[[include:ContentInstancesOfSpecificClasses.PropertyOverrides.Ruleset]]
```

![Example of using propertyOverrides attribute](./media/contentInstancesOfSpecificClasses-with-propertyOverrides-attribute.png)

### Attribute: `showImages`

> **Default value:** `false`

> Should image IDs be calculated for the returned instances. When `true`, [ImageIdOverride](../customization/ImageIdOverride.md) rules get applied when creating.

Reviewer: ```suggestion
Should image IDs be calculated for the returned instances. When `true`, [ImageIdOverride](../customization/ImageIdOverride.md) rules get applied when creating the content.
```

Also, would be nice to add a test with a snippet similar to [this](https://github.com/iTwin/itwinjs-core/blob/ea2f09b23710401e2f37a3c25af71b5e459304bd/full-stack-tests/presentation/src/learning-snippets/Hierarchies.test.ts#L1036-L1042).

Author: No change. Planned removal.

Reviewer: You're right, I completely forgot we were thinking of removing the `showImages` attribute, so there's no point to add a code snippet for it.

However, my suggested change for the description still applies.

### Attribute: `relatedInstances`

> Joining [related instances](../Common-Rules/RelatedInstanceSpecification.md) allows performing advanced filtering, grouping, labeling.

Reviewer: ```suggestion
Specifications of [related instances](../Common-Rules/RelatedInstanceSpecification.md) that can be used when creating the content. There are several use cases when this is useful:

- When there's a need to only load instances that have a related instance. Providing a [related instance](../Common-Rules/RelatedInstanceSpecification.md)
  specification with [isRequired](../Common-Rules/RelatedInstanceSpecification.md) set to `true` filters-out the instances that don't have the related instance.

- When there's a need to filter instances by a related instance value. The [alias](../Common-Rules/RelatedInstanceSpecification.md) attribute may then be used
  in the [`instanceFilter` attribute](#attribute-instancefilter) to reference related instance property values.

- When there's a need to customize content based on related instance property values. Related instance classes are included when looking for [customization rules](../Customization/index.md),
  which allows referencing related instances and their properties in [customization rule ECExpressions](../Customization/ECExpressions.md#override-value) by their
  [alias](../Common-Rules/RelatedInstanceSpecification.md).
```

```ts
[[include:ContentInstancesOfSpecificClasses.RelatedInstances.Ruleset]]
```

  |  | Result                                                                                                                                      |
  | ------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------- |

>   | SpatialViewDefinitions  | ![Example when doing normal property select](./media/contentInstancesOfSpecificClasses-with-relatedInstances-attribute-3.png) |
>   | ModelSelectors  | ![Example when doing normal property select](./media/contentInstancesOfSpecificClasses-with-relatedInstances-attribute-2.png) |
>   | ModelSelectors filtered by SpatialViewDefinition Yaw    | ![Example when selecting with related properties](./media/contentInstancesOfSpecificClasses-with-relatedInstances-attribute-1.png)     |

Reviewer: ```suggestion
  | `SpatialViewDefinition` instances  | ![A list of spatial view definitions](./media/contentInstancesOfSpecificClasses-with-relatedInstances-attribute-3.png) |
  | `ModelSelector` instances | ![A list of model selectors](./media/contentInstancesOfSpecificClasses-with-relatedInstances-attribute-2.png) |
  | `ModelSelector` instances filtered by `SpatialViewDefinition.Yaw`    | ![A list of model selectors filtered by yaw of related spatial view definition](./media/contentInstancesOfSpecificClasses-with-relatedInstances-attribute-1.png)     |
```

## Deprecated Attributes

### Attribute: `handleInstancesPolymorphically`

> **Default value:** `false`

> Defines default `classes` polymorphism value. If set to true, all target classes by default will include also include derived class instances. This value is overriden by the `classes`.[`arePolymorphic`](../Common-Rules/MultiSchemaClassesSpecification.md#attribute-arepolymorphic) value.

Reviewer: ```suggestion
Tells whether selecting instances from ECClasses specified in [`classes`](#attribute-classes) and [`excludedClasses`](#attribute-excludedclasses) attributes should be polymorphic or not.

The attribute was replaced by [MultiSchemaClasses.arePolymorphic](../Common-Rules/MultiSchemaClassesSpecification.md#attribute-arepolymorphic) attribute specified individually for each class definition under [`classes`](#attribute-classes) and [`excludedClasses`](#attribute-excludedclasses) attributes. At the moment, to keep backwards compatibility, this attribute acts as a fallback value in case the flag is not specified individually for a class definition.
```

### Attribute: `classes`

Defines a set of [multi schema classes](../Common-Rules/MultiSchemaClassesSpecification.md) that specify which ECClasses need to be selected to form the result.

```ts
[[include:ContentInstancesOfSpecificClasses.Classes.Ruleset]]
```

![Example of using "classes" attribute](./media/contentinstancesofspecificclasses-with-classes.png)

### Attribute: `excludedClasses`

> Defines a set of [multi schema classes](../Common-Rules/MultiSchemaClassesSpecification.md) that specify which ECClasses need to be excluded from the result.

Reviewer: I think it's worth mentioning that this is expected to define a subset of what's in `classes`.

```suggestion
Defines a set of [multi schema classes](../Common-Rules/MultiSchemaClassesSpecification.md) that specify which ECClasses need to be excluded from the result.

Since the classes included in the result are defined using the [`classes` attribute](#attribute-classes), it's expected that if defines a superset of `excludedClasses`. If not, `excludedClasses` has no effect.
```

Reviewer: I still feel like it would be worth adding a statement about the need for `excludedClasses` to derive from `classes` to have effect. @roluk, what do you think?

Reviewer 2: No need to be this specific, unless specifying unrelated classes exhibits some unexpected effect. I left my suggestion in another comment thread.

Reviewer: Thanks. The suggestion looks good.

```ts
[[include:ContentInstancesOfSpecificClasses.ExcludedClasses.Ruleset]]
```

  |                                       | Result                                                                                                                               |
  | ------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------ |
  | without excluded classes              | ![Example when doing normal class based instance select](./media/contentinstancesofspecificclasses-with-excludedclasses-1.png)       |
  | with `PhysicalModel` classes excluded | ![Example when selecting instances with some classes excluded](./media/contentinstancesofspecificclasses-with-excludedclasses-2.png) |

> 

Reviewer: extra new line

```ts
[[include:ContentInstancesOfSpecificClasses.RelatedProperties.Ruleset]]
```

  |         | Result                                                                                                                    |
  | ------- | ------------------------------------------------------------------------------------------------------------------------- |
  | without | ![Example when doing normal property select](./media/contentinstancesofspecificclasses-with-relatedproperties-1.png)      |
  | with    | ![Example when selecting with related properties](./media/contentinstancesofspecificclasses-with-relatedproperties-2.png) |

### Attribute: `calculatedProperties`

Specifications of [calculated properties](./CalculatedPropertiesSpecification.md) whose values are generated using provided [ECExpressions](../Advanced/ECExpressions.md#ecinstance).

```ts
[[include:ContentInstancesOfSpecificClasses.CalculatedProperties.Ruleset]]
```

![Example of using calculatedProperties attribute](./media/contentinstancesofspecificclasses-with-calculatedproperties.png)

### Attribute: `propertyCategories`

Specifications of [custom categories](PropertyCategorySpecification.md).

> Simply defining the categories does nothing - they have to be referenced through [`PropertySpecification.categoryId`](./PropertySpecification.md) specified in [`propertyOverrides`](#attribute-property-overrides) list.md) for more details.*

Reviewer: something's not right with the ending of this paragraph
